### PR TITLE
[master] Add websocket transport

### PR DESCRIPTION
**Pull request by dwoz**
### What does this PR do?

#64090

- Fix up tpc transport ssl options.
- Add the ability to define multiple verify_locations and set verify_flags via the ssl config setting
- Add websocket transport

### What issues does this PR fix or reference?
Fixes:

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by lukasraska** [comment]
What about https (e.g. wss:// scheme)? Also any plans to make this more efficient by having single WS multiplexed connection that unifies the currently separate publisher & request server, or is it meant to be "just a copy of TCP transport with WS instead"?

**Comment by dwoz** [comment]
> What about https (e.g. wss:// scheme)? Also any plans to make this more efficient by having single WS multiplexed connection that unifies the currently separate publisher & request server, or is it meant to be "just a copy of TCP transport with WS instead"?

We absolutely will be adding TLS support with both server and client certificates. As far as running both transports on the same port but different paths, yes the thought crossed my mind that we'll eventually go down that path. I'm trying to keep the option open.

File: salt/transport/ws.py [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+
+class PublishServer(salt.transport.base.DaemonizedPublishServer):
+    """ """
+
+    # TODO: opts!
+    # Based on default used in tornado.netutil.bind_sockets()
+    backlog = 128
```
Comment (dmurphy18) [inline salt/transport/ws.py:244]: wondering about use of Magic Number 128, what is it, and would it be better defined as some definition name set to 128 ?  Note it's use / setting in unchanged code, but wondering what it represents.

Cannot get the C/C++ and dislike of magic numbers out of me :smile:

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
Couple of spelling mistakes, and wonder what 128 for backlog represents

File: salt/transport/ws.py (hunk 1 above)
Comment (dwoz) [reply salt/transport/ws.py:244]: This represents the number of outstanding connections the socket will listen too and is a hold over from the TCP transport. This value has been in-place for some time. Unless it's causing problems I don't see the need to change it for now.

**Review by dmurphy18 — DISMISSED** [review]
Still need to resolve that Q about rel. vs absolute paths in doc/topics/transports/ssl.rst

File: doc/topics/transports/ssl.rst [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+option is passed as a dict and roughly corresponds to the options passed to the
+Python `ssl.wrap_socket <https://docs.python.org/3/library/ssl.html#ssl.wrap_socket>`_
+function for backwards compatability.
+
+.. versionadded:: 3007.0
+
+The ``ssl`` option accepts ``verify_locations`` and ``verify_flags``. The
+``verify_locations`` option is a list of strings or dictionaries. Strings are
+passed as a single argument to the SSL context's ``load_verify_locations``
+method. Dictionaries keys are expected to be one of ``cafile``, ``capath``,
```
Comment (twangboy) [inline doc/topics/transports/ssl.rst:21]: `Dictionary` keys...

File: doc/topics/transports/ssl.rst [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+``cadata``. For each corresponding key, the key and value will be passed as a
+keyword argument to ``load_verify_locations``. The ``verify_flags`` options is
```
Comment (twangboy) [inline doc/topics/transports/ssl.rst:23]: ... The \`\`verify_flags\`\` `option` is ...

File: doc/topics/transports/ssl.rst [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+A simple setup looks like this, on the Salt Master add the ``ssl`` option to the
+master configuration file:
+
+.. code-block:: yaml
+
+    ssl:
+      keyfile: <path_to_keyfile>
+      certfile: <path_to_certfile>
+
+A more complex  setup looks like this, on the Salt Master add the ``ssl``
```
Comment (twangboy) [inline doc/topics/transports/ssl.rst:36]: two spaces after `complex`

File: doc/topics/transports/ssl.rst [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+The minimal `ssl` option in the minion configuration file looks like this:
+
+.. code-block:: yaml
+
+    ssl: True
+    # Versions below 2016.11.4:
+    ssl: {}
+
+A Minion can be configured to present a client certificat to the master like this:
```
Comment (twangboy) [inline doc/topics/transports/ssl.rst:64]: `certificate`

File: doc/topics/transports/tcp.rst [hunk 6]
```diff
@@ -19,6 +19,14 @@ to ``tcp`` on each Salt minion and Salt master.
     use the same transport. We're investigating a report of an error when using
     mixed transport types at very heavy loads.
 
+
+TLS Support
+===========
+
+The TLS transport support full encryption and verification using both server
```
Comment (twangboy) [inline doc/topics/transports/tcp.rst:26]: supports

File: doc/topics/transports/ws.rst [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+===================
+
+The Websocket transport is an implementation of Salt's transport using the websocket protocol.
+The Websocket transport is enabled by changing the :conf_minion:`transport` setting
+to ``ws`` on each Salt minion and Salt master.
+
+TLS Support
+===========
+
+The Websocket transport support full encryption and verification using both server
```
Comment (twangboy) [inline doc/topics/transports/ws.rst:12]: supports

**Review by twangboy — CHANGES_REQUESTED** [review]

**Comment by dwoz** [comment]
> doc/topics/transports/ssl.rst

I added a note about it.

**Review by twangboy — DISMISSED** [review]

**Review by twangboy — APPROVED** [review]

**Comment by dwoz** [comment]
Merging. will address ssh test failures in a follow up PR.

**Pull request merged**